docs/source/howtos/add_model.rst

```diff
@@ -0,0 +1,406 @@
+            return {'e_init': self.default_e_init}
+
+        @property
+        def restrictions(self) -> dict[str, Any]:
+            return {'e_init': self.e_init_restrictions}
+
+
+How to add a new model class
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+With the data class in place, it is possible to create the model, which is a
+subclass of
+:py:class:`resolution_functions.models.model_base.InstrumentModel` (see its
+documentation for detailed specification of how to inherit from it). This
+**must** specify three class-level variables:
+
+* ``input`` - an integer specifying the number of arguments that the `
+  `__call__`` method takes
+
+* ``output`` - an integer specifying the number of outputs the ``__call__``
+  method returns
+
+* ``data_class`` - a reference to the data class
+  :ref:`created above<howto-model-dataclass>`, for example:
+
+.. code-block:: Python
+
+    class TestModel(InstrumentModel):
+        input = 1
+        output = 1
+        data_class = TestModelData
+
+
+Next, the ``__init__`` method should be defined. The function signature:
+
+* *Must* take ``model_data`` (an instance of the data class above) as its
+  first argument.
+* *May* take any number of other argument (usually representing
+  :term:`settings<setting>`, but may be others).
+* *Must not* expect the independent variables (i.e. the variables that the
+  model is a function of such as energy transfer or momentum).
+* *Must* take ``**kwargs``.
+
+The body of ``__init__()``:
+
+* *Must* call ``super().__init__``.
+* *Should* (if necessary) perform any validation of the arguments, e.g. that
+  the ``e_init`` is within the allowed range etc.
+* *Should* perform as much of the calculation as possible without the
+  independent variables. These pre-computed, intermediate values should be
+  stored as instance variables.
+
+  * For more complex calculations, it might be advisable to break them up into
+    multiple methods. Any such methods should be private (i.e. start with ``_``)
+    and, if possible, should be made ``@staticmethod`` or ``@classmethod``.
+
+* It *should not* keep a reference to ``model_data``.
+
+For example:
+
+.. code-block:: Python
+
+    from resolution_functions.models.model_base import InstrumentModel, InvalidInputError
+
+    class TestModel(InstrumentModel):
+        def __init__(self, model_data: TestModelData, e_init: float | None = None, **_):
+            super().__init__(model_data)
+
+            if e_init is None:
+                e_init = model_data.default_e_init
+            elif not model_data.e_init_restrictions[0] <= e_init <= model_data.e_init_restrictions[1]:
+                raise InvalidInputError('Good message')
+
+            self.useful_value = 0.5 * e_init ** 3
+
+Lastly, the ``__call__`` method must be implemented:
+
+* It *must* take as arguments all the independent variables that the model
+  models the :term:`resolution` as a function of.
+* It *must* accept ``*args`` and ``**kwargs``.
```

Comment:
Is this really a good idea? I suppose it is somewhat forced by formal OO interface type-checking things, but maybe there is a better way.

Anyway I guess it is accurate documentation for now 😅